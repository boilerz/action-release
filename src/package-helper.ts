import fs from 'fs';
import os from 'os';
import path from 'path';
import process from 'process';
import util from 'util';

import * as core from '@actions/core';
import * as exec from '@actions/exec';

const readdirAsync = fs.promises.readdir;
const readFileAsync = fs.promises.readFile;
const writeFileAsync = fs.promises.writeFile;
const existsAsync = util.promisify(fs.exists);

export enum Registry {
  NPM = 'https://registry.npmjs.org',
  GITHUB = 'https://npm.pkg.github.com',
}

interface PackageJson {
  version: string;
  devDependencies: Record<string, string>;
}

async function getPackageJson(): Promise<PackageJson> {
  const packageJsonPath = path.resolve(process.cwd(), 'package.json');
  core.debug(`ðŸ“¦ package.json path: ${packageJsonPath}`);

  const packageData: string = await readFileAsync(packageJsonPath, 'utf8');
  return JSON.parse(packageData);
}

export async function getCurrentVersion(): Promise<string> {
  const { version } = await getPackageJson();
  return version;
}

export async function getDevDependencies(): Promise<string[]> {
  const { devDependencies } = await getPackageJson();

  return Object.keys(devDependencies || {});
}

export async function readNpmRc(
  filePath: string,
): Promise<Record<string, string> | null> {
  if (!(await existsAsync(filePath))) return null;

  const fileRecord: Record<string, string> = {};
  const fileContent: string = await readFileAsync(filePath, 'utf8');
  const lines = fileContent.split(os.EOL).filter((line) => line !== '');
  lines.forEach((line) => {
    const [key, value] = line.split('=');
    fileRecord[key] = value;
  });

  return fileRecord;
}

export async function writeNpmRc(
  filePath: string,
  fileRecord: Record<string, string>,
): Promise<void> {
  const fileContent: string = Object.keys(fileRecord)
    .map((key) => `${key}=${fileRecord[key]}`)
    .join(os.EOL);
  await writeFileAsync(filePath, fileContent + os.EOL, { encoding: 'utf8' });
}

export function updateNpmRcForPublish(
  fileRecord: Record<string, string>,
): Record<string, string> {
  const updatedFileRecord = { ...fileRecord };
  Object.keys(fileRecord).forEach((key) => {
    if (/.*:?registry$/.test(key)) {
      delete updatedFileRecord[key];
    }
  });

  updatedFileRecord['always-auth'] = 'true';
  updatedFileRecord['//registry.npmjs.org/:_authToken'] = '${NPM_TOKEN}';
  updatedFileRecord['//npm.pkg.github.com/:_authToken'] = '${GITHUB_TOKEN}';

  return updatedFileRecord;
}

export async function setupNpmRcForPublish(): Promise<void> {
  const npmRcFilePath: string = path.resolve(process.cwd(), '.npmrc');
  const npmRcFileRecord: Record<string, string> =
    (await readNpmRc(npmRcFilePath)) || {};
  const updatedNpmRcFileRecord = await updateNpmRcForPublish(npmRcFileRecord);
  await writeNpmRc(npmRcFilePath, updatedNpmRcFileRecord);
}

export async function publish(
  registry: Registry,
  directory = '.',
): Promise<void> {
  core.debug(`ðŸ“’ Setting config registry at ${registry}`);
  await exec.exec('npm', ['config', 'set', 'registry', registry]);

  core.info(`ðŸ“¦ Publishing to ${registry}`);
  await exec.exec('npm', ['publish', directory, '--access', 'public']);
}

export async function listPackagePaths(folderPath: string): Promise<string[]> {
  const dirents = await readdirAsync(folderPath, { withFileTypes: true });
  if (dirents.some(({ name }) => name === 'package.json')) {
    return [folderPath];
  }
  const packagesPaths = await Promise.all(
    dirents
      .filter((dirent) => dirent.isDirectory())
      .map(({ name }) => listPackagePaths(path.resolve(folderPath, name))),
  );
  return Array.prototype.concat(...packagesPaths);
}
