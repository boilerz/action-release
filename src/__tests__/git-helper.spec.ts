import * as exec from '@actions/exec';
import * as github from '@actions/github';
import { GitHub } from '@actions/github/lib/utils';

import {
  Commit,
  createReleaseBody,
  detectBumpType,
  getCurrentBranch,
  release,
  version,
} from '../git-helper';
import * as packageHelper from '../package-helper';

function createCommit(message: string): Commit {
  return {
    id: 'bac6aee2d316d65025022f9e84f12eb2ffcb34ac',
    url: 'http://somewhere.web',
    message,
  } as Commit;
}

describe('git-helper', () => {
  describe('#getCurrentBranch', () => {
    it('should failed to detect branch without ref', () => {
      expect(
        getCurrentBranch.bind(null, null),
      ).toThrowErrorMatchingInlineSnapshot(`"Failed to detect branch"`);
    });

    it('should failed to detect branch without a well formed ref', () => {
      expect(
        getCurrentBranch.bind(null, ['']),
      ).toThrowErrorMatchingInlineSnapshot(
        `"Cannot retrieve branch name from GITHUB_REF"`,
      );
    });

    it('should detect branch successfully', () => {
      expect(getCurrentBranch('refs/heads/master')).toEqual('master');
    });
  });

  describe('#detectBumpType', () => {
    it('should failed to detect bump type with not enough commits', () => {
      expect(
        detectBumpType.bind(null, null),
      ).toThrowErrorMatchingInlineSnapshot(`"Failed to access commits"`);
      expect(detectBumpType.bind(null, [])).toThrowErrorMatchingInlineSnapshot(
        `"Failed to access commits"`,
      );
    });

    it('should return a minor bump type for features and minor message commit', () => {
      expect(
        detectBumpType([createCommit(':sparkles: feat something')]),
      ).toEqual('minor');
      expect(
        detectBumpType([createCommit(':sparkles: minor something else')]),
      ).toEqual('minor');
    });

    it('should return a patch bump type for any other commit type', () => {
      expect(detectBumpType([createCommit(':arrow_up: bump bar@1.0')])).toEqual(
        'patch',
      );
    });
  });

  describe('#version', () => {
    let execSpy: jest.SpyInstance;
    beforeEach(() => {
      execSpy = jest.spyOn(exec, 'exec').mockResolvedValue(0); // For safety
    });

    it('should version successfully', async () => {
      await version('minor');
      await version('minor', 'john@doe.co', 'jdoe');

      expect(execSpy.mock.calls).toMatchSnapshot();
    });
  });

  describe('#createReleaseBody', () => {
    it('should create a release body', () => {
      expect(
        createReleaseBody([
          createCommit('Merge 42'),
          createCommit(':arrow_up: Upgrade something'),
          createCommit(':arrow_up: Upgrade something else'),
          createCommit(':sparkles: New feature'),
          createCommit(':bug: fix issue'),
          createCommit('An other commit\ndetail1\ndetail2'),
        ]),
      ).toMatchInlineSnapshot(`
        "### :arrow_up: Dependencies update
        - Upgrade something ([bac6aee2](http://somewhere.web))
        - Upgrade something else ([bac6aee2](http://somewhere.web))

        ### :sparkles: Features
        - New feature ([bac6aee2](http://somewhere.web))

        ### :bug: Bug fixes
        - fix issue ([bac6aee2](http://somewhere.web))

        ### :card_file_box: Other changes
        - An other commit ([bac6aee2](http://somewhere.web))
          > detail1
          > detail2
        "
      `);

      expect(createReleaseBody([createCommit(':arrow_up: Upgrade something')]))
        .toMatchInlineSnapshot(`
        "### :arrow_up: Dependencies update
        - Upgrade something ([bac6aee2](http://somewhere.web))

        "
      `);
    });
  });

  describe('#release', () => {
    it('should fail to release without github token', async () => {
      await expect(
        release.bind(null, undefined),
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"Cannot release without GITHUB_TOKEN"`,
      );
    });

    it('should release successfully', async () => {
      const createReleaseSpy = jest.fn().mockResolvedValue({
        data: { id: '42' },
      });
      jest.spyOn(github, 'getOctokit').mockReturnValue(({
        repos: {
          createRelease: createReleaseSpy,
        },
      } as unknown) as InstanceType<typeof GitHub>);
      jest.spyOn(packageHelper, 'getCurrentVersion').mockResolvedValue('1.0.0');
      jest.spyOn(github.context, 'repo', 'get').mockReturnValue({
        owner: 'jdoe',
        repo: 'foo',
      });

      await release('github.token', [
        createCommit(':arrow_up: bump something to 1.0.0'),
      ]);

      expect(createReleaseSpy).toMatchInlineSnapshot(`
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "body": "### :arrow_up: Dependencies update
        - bump something to 1.0.0 ([bac6aee2](http://somewhere.web))

        ",
                "draft": false,
                "name": "1.0.0 (January 27, 2021)",
                "owner": "jdoe",
                "prerelease": false,
                "repo": "foo",
                "tag_name": "v1.0.0",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Promise {},
            },
          ],
        }
      `);
    });
  });
});
