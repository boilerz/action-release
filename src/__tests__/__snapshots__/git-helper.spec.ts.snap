// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git-helper #version should version successfully 1`] = `
Array [
  Array [
    "git",
    Array [
      "config",
      "user.name",
      "\\"jdoe\\"",
    ],
  ],
  Array [
    "git",
    Array [
      "config",
      "user.email",
      "\\"john@doe.co\\"",
    ],
  ],
  Array [
    "yarn",
    Array [
      "version",
      "--minor",
    ],
  ],
  Array [
    "git",
    Array [
      "status",
      "-uno",
    ],
    Object {
      "listeners": Object {
        "stdout": [Function],
      },
    },
  ],
  Array [
    "git",
    Array [
      "push",
    ],
  ],
  Array [
    "git",
    Array [
      "push",
      "--tags",
    ],
  ],
  Array [
    "git",
    Array [
      "config",
      "user.name",
      "\\"jdoe\\"",
    ],
  ],
  Array [
    "git",
    Array [
      "config",
      "user.email",
      "\\"john@doe.co\\"",
    ],
  ],
  Array [
    "yarn",
    Array [
      "version",
      "--minor",
    ],
  ],
  Array [
    "git",
    Array [
      "status",
      "-uno",
    ],
    Object {
      "listeners": Object {
        "stdout": [Function],
      },
    },
  ],
  Array [
    "git",
    Array [
      "push",
    ],
  ],
  Array [
    "git",
    Array [
      "push",
      "--tags",
    ],
  ],
]
`;
